libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,120 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+#include <deque>
+#include <memory>
+#include <ostream>
+#include <variant>
+#include <vector>
+
+namespace solidity::frontend
+{
+
+/**
+ * Creates a Function call graph for a contract at the granularity of Solidity
+ * functions and modifiers. The graph can represent the situation either at contract creation
+ * or at deployment time.
+ *
+ * Includes the following special nodes:
+ *  - Entry: represents a call from the outside of the contract.
+ *    At deployment this is the node that connects to all the functions exposed through the external
+ *    interface. At contract creation it connects to the constructor and variable initializers in
+ *    the bottom-most contract in the linearized inheritance order.
+ *  - InternalDispatch: Represents the internal dispatch function, which calls internal functions
+ *    determined at runtime by values of variables and expressions. Functions that are not called
+ *    right away get an edge from this node.
+ *
+ *  Nodes are a variant of either the enum SpecialNode or a CallableDeclaration which currently
+ *  can be a function or a modifier. There are no nodes representing event calls. Instead all
+ *  emitted events and created contracts are gathered in separate sets included in the graph just
+ *  for that purpose.
+ *
+ *  Auto-generated getter functions for public state variables are ignored.
+ *
+ *  Only calls reachable from an Entry node are included in the graph.
+ */
```

Comment:
This originally mentioned that there could be nodes other than callables - e.g. contracts to represent implicit constructors. Or that function calls without a definition could be represented by `FunctionCall` nodes.

None of that was implemented so I removed that from the docstring. I have also changed `Node` to hold `CallableDeclaration` rather than `ASTNode`. If we want to add more node types, I'd rather just add them as separate variants and not have an overly generic pointer that has to be always cast and asserted against.

Anyway, do we want any of that in this PR or it it fine the way it is now?

Reply:
Looks good!